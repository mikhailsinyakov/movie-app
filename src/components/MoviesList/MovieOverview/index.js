import React from "react";
import PropTypes from "prop-types";
import { useHistory } from "react-router-dom";
import styled from "styled-components";
import { isNull } from "../../../helpers/customPropCheckers";
import Poster from "../../../shared/Poster";
import Info from "./Info";

const MovieOverview = ({ movie, className }) => {
  const history = useHistory();
  return (
    <div
      className={className}
      onClick={() => history.push(`/movie/${movie.id}`)}
    >
      <Poster source={movie.poster_src} title={movie.title} />
      <Info
        title={movie.title}
        rating={movie.vote_average}
        genres={movie.genres}
      />
    </div>
  );
};

MovieOverview.propTypes = {
  movie: PropTypes.shape({
    title: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    poster_src: PropTypes.oneOfType([PropTypes.string, isNull]),
    vote_average: PropTypes.number,
    genres: PropTypes.arrayOf(PropTypes.string)
  })
};

const StyledMovieOverview = styled(MovieOverview)`
  width: var(--poster-width);
  margin: 1rem;
  border-radius: 1rem;
  background-color: #442c2c;
  color: #9a9a86;
  height: fit-content;
  cursor: pointer;
  transition: transform 0.5s;

  &:hover {
    transform: translateY(-1rem);
  }
`;

export default StyledMovieOverview;
