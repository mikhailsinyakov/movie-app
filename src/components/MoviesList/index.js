import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { withLanguageContext } from "../Language";
import MovieOverview from "./MovieOverview";
import Button from "shared/Button";
import { isNull } from "helpers/customPropCheckers";

const Loading = styled.div`
  text-align: center;
  text-shadow: 0 0 10px black;
  color: #ded6ca;
  padding: 3rem;
`;

const Wrapper = styled.div`
  text-align: center;
  padding-bottom: 1rem;
  max-width: 1000px;
  margin: 0 auto;
`;

const MoviesList = ({ getUIText, moviesData, getMoreMovies, className }) => {
  const { list, loading, completed } = moviesData;

  return (
    <Wrapper>
      <div className={className}>
        {list.map(movie => (
          <MovieOverview key={movie.id} movie={movie} />
        ))}
      </div>
      {loading ? (
        <Loading>{getUIText("loading")}...</Loading>
      ) : !completed ? (
        <Button onClick={getMoreMovies}>{getUIText("more")}</Button>
      ) : null}
    </Wrapper>
  );
};

MoviesList.propTypes = {
  getUIText: PropTypes.func.isRequired,
  moviesData: PropTypes.shape({
    list: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.string.isRequired,
        id: PropTypes.number.isRequired,
        poster_src: PropTypes.oneOfType([PropTypes.string, isNull]),
        vote_average: PropTypes.number,
        genres: PropTypes.arrayOf(PropTypes.string)
      })
    ).isRequired,
    loading: PropTypes.bool.isRequired,
    completed: PropTypes.bool.isRequired
  }).isRequired,
  getMoreMovies: PropTypes.func.isRequired
};

const StyledMoviesList = styled(MoviesList)`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
`;

export default withLanguageContext(StyledMoviesList);
