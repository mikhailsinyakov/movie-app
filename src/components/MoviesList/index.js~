import React, { useState, useEffect, useRef } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { withLanguageContext } from "../Language";
import MovieOverview from "./MovieOverview";
import Button from "shared/Button";
import Error from "components/Error";
import { getMoviesList } from "api/movieAPI";
import cache from "cache";

const Wrapper = styled.div`
  text-align: center;
  padding-bottom: 1rem;
  max-width: 1000px;
  margin: 0 auto;
`;

const MoviesList = ({ language, getUIText, className }) => {
  const isMounted = useRef(false);
  const [moviesData, setMoviesData] = useState({
    page: 0,
    total_pages: 0,
    results: []
  });

  const [error, setError] = useState(null);
  
  useEffect(() => {
    window.document.title = getUIText("appName");
  }, [language]);

  useEffect(() => {
    isMounted.current = true;
    return () => (isMounted.current = false);
  }, []);

  useEffect(() => {
    let ignore = false;
    if (cache.has("movies", "latest", language)) {
      setMoviesData(cache.get("movies", "latest", language));
    } else {
      getMoviesList({ language })
        .then(initMoviesData => {
          cache.set("movies", "latest", language, initMoviesData);
          !ignore && setMoviesData(initMoviesData);
        })
        .catch(setError);
    }
    return () => (ignore = true);
  }, [language]);

  const addMoviesData = () => {
    getMoviesList({ language, page: moviesData.page + 1 })
      .then(newMoviesData => {
        newMoviesData.results = [
          ...moviesData.results,
          ...newMoviesData.results
        ];
        cache.set("movies", "latest", language, newMoviesData);
        isMounted.current && setMoviesData(newMoviesData);
      })
      .catch(setError);
  };

  return (
    <Wrapper>
      <div className={className}>
        {moviesData.results.map(movie => (
          <MovieOverview key={movie.id} movie={movie} />
        ))}
      </div>
      {error && <Error />}
      {moviesData.page !== moviesData.total_pages && !error ? (
        <Button onClick={addMoviesData}>{getUIText("more")}</Button>
      ) : null}
    </Wrapper>
  );
};

MoviesList.propTypes = {
  language: PropTypes.string.isRequired,
  getUIText: PropTypes.func.isRequired
};

const StyledMoviesList = styled(MoviesList)`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
`;

export default withLanguageContext(StyledMoviesList);
