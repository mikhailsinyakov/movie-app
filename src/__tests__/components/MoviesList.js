import React from "react";
import Renderer from "react-test-renderer";
import { MemoryRouter as Router } from "react-router-dom";
import { fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import withLanguage from "../../components/Language";
import MoviesList from "../../components/MoviesList";

const Parent = ({ children }) => <div>{children}</div>;
const ParentWithLanguage = withLanguage(Parent);

let moviesData = {
  list: [
    {
      title: "Fight club",
      id: 555,
      poster_src: "poster.jpg",
      vote_average: 8.9,
      genres: ["Action", "Drama"]
    },
    {
      title: "Amiley",
      id: 123,
      poster_src: "poster2.jpg",
      vote_average: 7.5,
      genres: ["Drama"]
    },
    {
      title: "Armageddon",
      id: 588,
      poster_src: "poster3.jpg",
      vote_average: 8.4,
      genres: ["Action"]
    }
  ],
  loading: false,
  completed: false
};
const getMoreMovies = jest.fn();

it("matches snapshot", () => {
  const component = Renderer.create(
    <Router>
      <ParentWithLanguage>
        <MoviesList moviesData={moviesData} getMoreMovies={getMoreMovies} />
      </ParentWithLanguage>
    </Router>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

const renderComponent = moviesData =>
  render(
    <Router>
      <ParentWithLanguage>
        <MoviesList moviesData={moviesData} getMoreMovies={getMoreMovies} />
      </ParentWithLanguage>
    </Router>
  );

describe("with loading = false and completed = false", () => {
  it("renders an element with 'More' text", () => {
    const { getByText } = renderComponent(moviesData);
    expect(getByText("More")).toBeInTheDocument();
  });

  it("renders no element with 'Loading' text", () => {
    const { queryByText } = renderComponent(moviesData);
    expect(queryByText("Loading...")).toBeNull();
  });

  it("renders 3 images", () => {
    const { container } = renderComponent(moviesData);
    expect(container.getElementsByTagName("img")).toHaveLength(3);
  });

  it("renders first movie image, title, rating, and genres", () => {
    const { getByText, getAllByText, getByAltText } = renderComponent(
      moviesData
    );
    const movieInfo = moviesData.list[0];
    expect(getByAltText(movieInfo.title)).toBeInTheDocument();
    expect(getByText(movieInfo.title)).toBeInTheDocument();
    expect(getByText(movieInfo.vote_average.toString())).toBeInTheDocument();
    expect(getAllByText(movieInfo.genres[0])).toBeTruthy();
  });

  it("calls getMoreMovies function when click 'More' button", () => {
    const { getByText } = renderComponent(moviesData);
    fireEvent.click(getByText("More"));
    expect(getMoreMovies).toHaveBeenCalled();
  });
});

describe("with loading = true and completed = false", () => {
  beforeAll(() => {
    moviesData.loading = true;
    moviesData.completed = false;
  });

  it("renders 'Loading...' text", () => {
    const { getByText } = renderComponent(moviesData);
    expect(getByText("Loading...")).toBeInTheDocument();
  });

  it("renders no button with 'More' text", () => {
    const { queryByText } = renderComponent(moviesData);
    expect(queryByText("More")).toBeNull();
  });
});

describe("with loading = false and completed = true", () => {
  beforeAll(() => {
    moviesData.loading = false;
    moviesData.completed = true;
  });

  it("renders no element with 'Loading...' text", () => {
    const { queryByText } = renderComponent(moviesData);
    expect(queryByText("Loading...")).toBeNull();
  });

  it("renders no button with 'More' text", () => {
    const { queryByText } = renderComponent(moviesData);
    expect(queryByText("More")).toBeNull();
  });
});
