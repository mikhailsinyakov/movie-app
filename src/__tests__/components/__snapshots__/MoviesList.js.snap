// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <main>
    <div>
      <img
        alt="Fight club"
        src="poster.jpg"
      />
      <div>
        <p>
          Fight club
          <span>
            8.9
          </span>
        </p>
        <p>
          <span>
            Action
          </span>
          <span>
            Drama
          </span>
        </p>
      </div>
    </div>
    <div>
      <img
        alt="Amiley"
        src="poster2.jpg"
      />
      <div>
        <p>
          Amiley
          <span>
            7.5
          </span>
        </p>
        <p>
          <span>
            Drama
          </span>
        </p>
      </div>
    </div>
    <div>
      <img
        alt="Armageddon"
        src="poster3.jpg"
      />
      <div>
        <p>
          Armageddon
          <span>
            8.4
          </span>
        </p>
        <p>
          <span>
            Action
          </span>
        </p>
      </div>
    </div>
    <button
      onClick={[MockFunction]}
    >
      More
    </button>
  </main>
</div>
`;
